### ISSUE TITLE
BUG: Some DTypes have identical hashes but distinct ids

### ISSUE THREAD

**Issue by ngoldbaum** [numpy/numpy#26446]
Matti originally found this on a 32 bit system, but on 64 bit:

```
>>> hash(np.dtype('uint64'))
-3444929848325097739
>>> hash(np.dtype('ulonglong'))
-3444929848325097739
>>> np.dtype('uint64') is np.dtype('ulonglong')
False
>>> id(np.dtype('uint64')), id(np.dtype('ulonglong'))
(4424700496, 4424700736)
```

I'm guessing this is a problem with the way we're creating these aliases at a low level in C. Somehow python exposes distinct objects, but `tp_hash` sees the exact same object.

_Originally posted by @mattip in https://github.com/numpy/numpy/issues/26439#issuecomment-2112882590_

**Comment by seberg** [comment]
They are equal, no?  So the hashes should be identical. There is a lot wrong with hashes in dtypes but I I would think this is right?

**Comment by MegaIng** [comment]
It is *required* that `tp_hash` and `tp_richcompare` are consist with each other (i.e. `a == b` implies `hash(a) == hash(b)`), otherwise the hash implementation is broken.

**Comment by ngoldbaum** [comment]
Sorry, you're right of course. My intuition is thrown off because these are distinct types, which I expect to have distinct hashes, but you're right the rules about hashing and equality forces this.

I guess this is what's inconsistent then:

```
>>> np.uint64 == np.ulonglong
False
```

Anyway, not a bug, closing and sorry for the noise.

**Comment by MegaIng** [comment]
There is actually a bug here it appears: https://discuss.python.org/t/how-does-cpython-calculate-the-hashes-of-types/53438/6

```
>>> np.float64 == np.dtype("float64")
True
>>> hash(np.float64) == hash(np.dtype("float64"))
False
```

This is a broken hash implementation that needs to be fixed, otherwise `dict` and `set` with these objects might not work correctly.

**Comment by ngoldbaum** [comment]
I think this is part of "There is a lot wrong with hashes in dtypes" that Sebastian was referring to above. `np.float64` is a scalar type and `np.dtype('float64')` is a dtype instance. They're distinct types and objects but numpy likes to pretend they aren't, leading to all kinds of issues. Fixing this would likely break a lot of code that assumes dtypes and scalars are interchangeable.

**Comment by rkern** [comment]
Yeah, that's a known issue. We overload `dtype.__eq__` to be too promiscuous such that it is literally impossible to satisfy the "if they compare equal, they hash equal" rule, if we consider the universe of all possible Python objects. _But_, it does hold true for more restricted universes of objects that one actually usually wants to use as keys in the same `dict`. So we leave it alone for now.

**Comment by mattip** [comment]
The problem came up on 32-bit linux where either `isdtype(np.int32, 'integral')` or `isdtype(np.intc, 'integral')` returned `False`. The cause of this bug was the inconsistency between these `set()` calls:
```
>>> set([np.dtype("int32"), np.dtype("intc")])
{dtype('int32')}  # sometimes could also yield `intc`
>>> set([np.dtype("int32").type, np.dtype("intc").type])
{<class 'numpy.intc'>, <class 'numpy.int32'>}
```

This was fixed in #26441

**Comment by seberg** [comment]
One day, I think we should make `type(dtype("long")) == type(dtype("int"))` and I suspect we can get away with it without too many problems.
(Which is nice, because that is the part that ufuncs care more about now.)

I am slightly more on edge about trying to get rid of these "not quite aliases" for scalars and dtype instances because it would mean that `np.dtype("l").char` cannot always round-trip, which seems very confusing.
